pkgs/development/python-modules/nassl/default.nix

```diff
@@ -0,0 +1,120 @@
+{ lib
+, fetchFromGitHub
+, fetchurl
+, buildPythonPackage
+, pkgsStatic
+, openssl
+, invoke
+, pytest
+, tls-parser
+, cacert
+}:
+
+let
+  zlibStatic = pkgsStatic.zlib;
+  nasslOpensslArgs = {
+    static = true;
+    enableSSL2 = true;
+  };
+  nasslOpensslFlagsCommon = [
+    "zlib"
+    "no-zlib-dynamic"
+    "no-shared"
+    "--with-zlib-lib=${zlibStatic.out}/lib"
+    "--with-zlib-include=${zlibStatic.out.dev}/include"
+    "enable-rc5"
+    "enable-md2"
+    "enable-gost"
+    "enable-cast"
+    "enable-idea"
+    "enable-ripemd"
+    "enable-mdc2"
+    "-fPIC"
+  ];
+  opensslStatic = (openssl.override nasslOpensslArgs).overrideAttrs (
+    oldAttrs: rec {
+      name = "openssl-${version}";
+      version = "1.1.1";
+      src = fetchurl {
+        url = "https://www.openssl.org/source/${name}.tar.gz";
+        sha256 = "0gbab2fjgms1kx5xjvqx8bxhr98k4r8l2fa8vw7kvh491xd8fdi8";
+      };
+      configureFlags = oldAttrs.configureFlags ++ nasslOpensslFlagsCommon ++ [
+        "enable-weak-ssl-ciphers"
+        "enable-tls1_3"
+        "no-async"
+      ];
+      patches = [ ./nix-ssl-cert-file.patch ];
+      buildInputs = oldAttrs.buildInputs ++ [ zlibStatic cacert ];
+    }
+  );
+  opensslLegacyStatic = (openssl.override nasslOpensslArgs).overrideAttrs (
+    oldAttrs: rec {
+      name = "openssl-${version}";
+      version = "1.0.2e";
+      src = fetchurl {
+        url = "https://www.openssl.org/source/${name}.tar.gz";
+        sha256 = "1zqb1rff1wikc62a7vj5qxd1k191m8qif5d05mwdxz2wnzywlg72";
+      };
+      configureFlags = oldAttrs.configureFlags ++ nasslOpensslFlagsCommon;
+      patches = [ ];
+      buildInputs = oldAttrs.buildInputs ++ [ zlibStatic ];
+      # openssl_1_0_2 needs `withDocs = false`
+      outputs = lib.remove "doc" oldAttrs.outputs;
+    }
+  );
```

Comment:
This all seems to be pretty fragile to me, and I'm not a big fan of continuing usage of openssl1.0 which has has known CVEs open.

opinions? @jtojnar @mweinelt @vcunat

Reply:
Would it be possible to patch the library to link against system openssl instead of building a local static copy? Maybe open an issue upstream.

Reply:
The downside of missing this is not being able to scan for things that were removed (SSLv3?) from more recent OpenSSL versions AFAIUI. I'm not sure there is an actual need to scan for legacy stuff, but others might disagree.

I've looked into the `testssl.sh` package and it is shipped without a legacy OpenSSL package.

Reply:
Thanks for your comments! 

I've already tried to 

1. build Nassl without the legacy OpenSSL version potentially accepting a lack of functionality. Unfortunately, there's no flag and the code depends on the legacy version on quite a few places.

2. use the Nixpkgs OpenSSL version (i.e., at the time of writing, `1.1.1g`). With this version, however, Nassl tests started to fail.

AFAICS, Nassl depends on these specific versions of both the legacy and modern OpenSSL library.

Reply:
Could you try opening an issue upstream about linking against system openssl?

Reply:
Sure! Here you go: https://github.com/nabla-c0d3/nassl/issues/64

Reply:
By that thread it seems the 1.0.2 security issues won't be a problem in here.

Reply:
For 1.1 it also can't use our nixpkgs-wide build?  (perhaps due to missing `"enable-weak-ssl-ciphers"`)

Reply:
I've tried that and, unfortunately, tests started to fail.

NB: I've just updated `tls-parser` and `sslyze` to the latest upstream version.